compose/ui/ui/src/desktopMain/kotlin/androidx/compose/ui/window/WindowDecoration.desktop.kt

```diff
@@ -0,0 +1,70 @@
+/*
+ * Copyright 2024 The Android Open Source Project
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package androidx.compose.ui.window
+
+import androidx.compose.runtime.Immutable
+import androidx.compose.ui.unit.Dp
+import androidx.compose.ui.unit.dp
+
+/**
+ * Defines the options for window decoration.
+ */
+interface WindowDecoration {
```

Comment:
```suggestion
sealed class WindowDecoration {
```

AOSP uses `sealed class` in this case, it isn't possible to have custom implementation.

Reply:
I thought we didn't want to use `sealed` or `enum` classes because it can make it hard to provide backwards compatibility later when we want to add additional subtypes. For example, someone can write
```
when (decoration) {
    is System -> ...
    is Undecorated -> ...
}
```
and when we add another type, this will be broken.

Reply:
Right 🤔, let's keep `interface`

Reply:
Agree with @igordmn - it should be `sealed`, but not `class`, `sealed interface` should work here. I guess it's covered by some policy - need to look for formal rule there, but the idea is to avoid implementation outside of our code. Explicit requirement in case of adding is pro, not cons

Reply:
> Agree with @igordmn - it should be `sealed`, but not `class`, `sealed interface` should work here. I guess it's covered by some policy - need to look for formal rule there, but the idea is to avoid implementation outside of our code. Explicit requirement in case of adding is pro, not cons

If we design this interface similar to `TextFieldDecoration `and move the implementation of the `Undecorated Resizer` to this part, it can be provided to users to achieve an immersive title bar application experience on Windows. This requires exposing the `Window `to set window properties and exposing the `HitTest `related interfaces for hit testing of controls within the title bar. You can refer to this [existing implementation](https://github.com/Konyaco/compose-fluent-ui/pull/57/commits/b5f6aa3888d6b6f6295e0569ea0355bd9c09d50b) for specifics. Additionally, in the current Resizer implementation, if the user calls a `Dialog `or `Popup`, the Resizer response will be blocked and won’t respond . If we implement the Resizer through this new interface and always place the `WindowDecoration `field above the `ContentLayer`, can this solve the problem? Would this design make the Decoration API more user-friendly and extensible?